Panic from repricing LP orders

Hitting a panic when forwarding time on null chain due to exotic LP orders and their repricing

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1038bc33c]

goroutine 49820 [running]:
code.vegaprotocol.io/vega/core/execution.(*Market).applyOffset(0x1400182f440, 0x1, 0x0, 0x14011672980)
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/liquidity_provision.go:525 +0x9c
code.vegaprotocol.io/vega/core/execution.(*Market).adjustPrice(0x1401c65da58?, 0x1, 0x14000372000?, 0xa?, 0x1401c65da48?, 0x102afdaf8?)
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/liquidity_provision.go:507 +0x2c
code.vegaprotocol.io/vega/core/execution.(*Market).repriceLiquidityOrder(0x1400182f440, 0x5eb0d20?, 0xb7, 0x1?)
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/liquidity_provision.go:503 +0x120
code.vegaprotocol.io/vega/core/liquidity.(*Engine).createOrUpdateForParty(0x1400d9064d0, {0x105fbc4c0, 0x14015181ad0}, 0x14011b52dc0?, 0x1400e9985a0?, {0x1400ec60880, 0x40}, 0x1400e6ee048)
        /Users/tom/Code/vega-market-sim/extern/vega/core/liquidity/engine.go:583 +0x25c
code.vegaprotocol.io/vega/core/liquidity.(*Engine).Update(0x1400d9064d0, {0x105fbc4c0, 0x14015181ad0}, 0x2?, 0x40?, 0x1400e531c40?)
        /Users/tom/Code/vega-market-sim/extern/vega/core/liquidity/engine.go:541 +0xe0
code.vegaprotocol.io/vega/core/execution.(*Market).repriceAllSpecialOrders(0x1400182f440, {0x105fbc4c0, 0x14015181ad0}, 0x8?, {0x14015041a30?, 0x105ed16a0?, 0x1401c65e288?}, 0x102906fb0?, 0x1401c65e268?)
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/special_orders.go:194 +0x394
code.vegaprotocol.io/vega/core/execution.(*Market).checkForReferenceMoves(0x1400182f440, {0x105fbc4c0, 0x14015181ad0}, {0x14015041a30, 0x2, 0x2}, 0x1)
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/reference_price_moves.go:80 +0x2b4
code.vegaprotocol.io/vega/core/execution.(*Market).leaveAuction(0x1400182f440, {0x105fbc4c0, 0x14015181ad0}, {0x14018b849c0?, 0x1?, 0x0?})
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/market.go:1114 +0x890
code.vegaprotocol.io/vega/core/execution.(*Market).checkAuction(0x1400182f440, {0x105fbc4c0, 0x14015181ad0}, {0xf710890?, 0xed936bd75?, 0x0?})
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/auction.go:83 +0x494
code.vegaprotocol.io/vega/core/execution.(*Market).OnTick(0x1400182f440, {0x105fbc4c0, 0x14015181ad0}, {0x0?, 0x0?, 0x0?})
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/market.go:731 +0x5e4
code.vegaprotocol.io/vega/core/execution.(*Engine).OnTick(0x14000465880, {0x105fbc4c0, 0x14015181ad0}, {0x1?, 0x105fd6098?, 0x0?})
        /Users/tom/Code/vega-market-sim/extern/vega/core/execution/engine.go:778 +0x1a0
code.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)
        /Users/tom/Code/vega-market-sim/extern/vega/core/vegatime/service.go:104
code.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0x140003d6600, {0x105fbc4c0, 0x14015181ad0}, {0x107a79760?, 0x105817a60?, 0x0?})
        /Users/tom/Code/vega-market-sim/extern/vega/core/vegatime/service.go:77 +0x278
code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0x1400d949100, 0x20, 0x20}, {{0x0, 0x0}, {0x14002564338, 0x6}, 0xe29, {0xf710890, ...}, ...}, ...})
        /Users/tom/Code/vega-market-sim/extern/vega/core/processor/abci.go:781 +0x554
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0x1400d949100, 0x20, 0x20}, {{0x0, 0x0}, {0x14002564338, 0x6}, 0xe29, {0xf710890, ...}, ...}, ...})
        /Users/tom/Code/vega-market-sim/extern/vega/core/blockchain/abci/abci.go:46 +0x60
code.vegaprotocol.io/vega/cmd/vega/node.(*appW).BeginBlock(_, {{0x1400d949100, 0x20, 0x20}, {{0x0, 0x0}, {0x14002564338, 0x6}, 0xe29, {0xf710890, ...}, ...}, ...})
        /Users/tom/Code/vega-market-sim/extern/vega/cmd/vega/node/app_wrapper.go:64 +0x4c
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).BeginBlock(0x140001f4480)
        /Users/tom/Code/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:336 +0x2ac
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).processBlock(0x140001f4480)
        /Users/tom/Code/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:195 +0x1b0
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).ForwardTime(0x140001f4480, 0x3b9aca00)
        /Users/tom/Code/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:272 +0x198
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime.func1()
        /Users/tom/Code/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:129 +0x2c
created by code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime
        /Users/tom/Code/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:128 +0x1ec
```

[replay.zip](https://github.com/vegaprotocol/vega/files/10958234/replay.zip)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -61,6 +61,7 @@
 - [7805](https://github.com/vegaprotocol/vega/issues/7805) - Fix re-announcing node in the same epoch kills data node.
 - [7820](https://github.com/vegaprotocol/vega/issues/7820) - Remove the check for past date in limits engine
 - [7822](https://github.com/vegaprotocol/vega/issues/7822) - Fix get last epoch query
+- [7823](https://github.com/vegaprotocol/vega/issues/7823) - Fix validation of liquidity provisions shapes references
 
 ## 0.68.0
 

--- a/commands/liquidity_provision_submission.go
+++ b/commands/liquidity_provision_submission.go
@@ -95,6 +95,12 @@ func checkLiquidityProvisionShape(
 	zero := big.NewInt(0)
 
 	for idx, order := range orders {
+		if _, ok := types.PeggedReference_name[int32(order.Reference)]; !ok {
+			errs.AddForProperty(
+				fmt.Sprintf("%v.%d.reference", shapeSideField, idx),
+				ErrIsNotValid,
+			)
+		}
 		if order.Reference == types.PeggedReference_PEGGED_REFERENCE_UNSPECIFIED {
 			errs.AddForProperty(
 				fmt.Sprintf("%v.%d.reference", shapeSideField, idx),

--- a/commands/liquidity_provision_submission_test.go
+++ b/commands/liquidity_provision_submission_test.go
@@ -170,6 +170,20 @@ func TestLiquidityProvisionSubmission(t *testing.T) {
 			lp:        commandspb.LiquidityProvisionSubmission{},
 			errString: "liquidity_provision_submission.buys (empty shape), liquidity_provision_submission.commitment_amount (is required), liquidity_provision_submission.fee (is required), liquidity_provision_submission.market_id (is required), liquidity_provision_submission.sells (empty shape)",
 		},
+		{
+			lp: commandspb.LiquidityProvisionSubmission{
+				CommitmentAmount: "100",
+				Fee:              "0.1",
+				MarketId:         "08dce6ebf50e34fedee32860b6f459824e4b834762ea66a96504fdc57a9c4741",
+				Sells: []*types.LiquidityOrder{
+					{Reference: types.PeggedReference_PEGGED_REFERENCE_BEST_ASK, Offset: "0", Proportion: 1},
+				},
+				Buys: []*types.LiquidityOrder{
+					{Reference: types.PeggedReference(186), Offset: "0", Proportion: 1},
+				},
+			},
+			errString: "liquidity_provision_submission.buys.0.reference (is not a valid value)",
+		},
 	}
 
 	for i, c := range cases {

